### DISCUSSION TITLE
Type issue with GCG Typegen for Graphcache config

### DISCUSSION THREAD

**Discussion by disbelief** [urql-graphql/urql#1654] [Q&A] [answered]
Hi there,

Maybe I'm trying to use the new GCG typegen feature too early, but I can't seem to get it working.

I've followed the instructions @JoviDeCroock provided to generate a config for the Graphcache exchange: https://github.com/FormidableLabs/urql/issues/901#issuecomment-840521982

But when I try to use the generated config, I'm seeing a Typescript error and my code won't compile.

Here's the specific error:
```
TypeScript error in /Users/mlawlor/Projects/Dusk/code/faction/src/hooks/useClient.ts(135,27):
Type 'GraphCacheConfig' does not satisfy the constraint 'Partial<CacheExchangeOpts>'.
  Types of property 'updates' are incompatible.
    Type 'GraphCacheUpdaters | undefined' is not assignable to type 'Partial<UpdatesConfig> | undefined'.
      Type 'GraphCacheUpdaters' is not assignable to type 'Partial<UpdatesConfig>'.
        Types of property 'Mutation' are incompatible.
          Type '{ acceptInvite?: UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs> | undefined; ... 41 more ...; updateUser?: UpdateResolver<...> | undefined; } | undefined' is not assignable to type '{ [fieldName: string]: UpdateResolver<DataFields, Variables>; } | undefined'.
            Type '{ acceptInvite?: UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs> | undefined; ... 41 more ...; updateUser?: UpdateResolver<...> | undefined; }' is not assignable to type '{ [fieldName: string]: UpdateResolver<DataFields, Variables>; }'.
              Property 'acceptInvite' is incompatible with index signature.
                Type 'UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs>' is not assignable to type 'UpdateResolver<DataFields, Variables>'.
                  Property 'acceptInvite' is missing in type 'DataFields' but required in type '{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }'.  TS2344

    133 |             devtoolsExchange,
    134 |             dedupExchange,
  > 135 |             cacheExchange<GraphCacheConfig>({
        |                           ^
```

Here's how I initialize the exchange:

```typescript
import { GraphCacheConfig } from '__generated__/types';
import schema from '__generated__/schema';
import updates from './updates';

const client = createClient({
  url: process.env.REACT_APP_API_URL || '/graphql',
  requestPolicy: 'cache-first',
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange<GraphCacheConfig>({
      schema,
      updates,
    }),
    authExchange<ClientStateType>({
      getAuth,
      addAuthToOperation,
      didAuthError,
      willAuthError,
    }),
    fetchExchange,
  ],
});
```

The error is present regardless if I call `cacheExchange<GrapheCacheConfig>()` with or without arguments. It seems to be the generated `GraphCacheUpdaters` type that it doesn't like.

Here are the lines of the generated GraphCache config that the error seems to be referencing:

```typescript
export type GraphCacheUpdaters = {
  Mutation?: {
    acceptInvite?: GraphCacheUpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>> }, MutationAcceptInviteArgs>,
    // ... more update resolvers with similar signatures
  },
  Subscription?: {},
};

export type GraphCacheConfig = {
  updates?: GraphCacheUpdaters,
  keys?: GraphCacheKeysConfig,
  optimistic?: GraphCacheOptimisticUpdaters,
  resolvers?: GraphCacheResolvers
};
```

Here is my codegen.yml:
```yaml
overwrite: true
schema: 'http://localhost:4000/graphql'
generates:
  __generated__/types.ts:
    plugins:
      - typescript
      - typescript-urql-graphcache
  __generated__/schema.ts:
    plugins:
      - 'urql-introspection'
    config:
      useTypeImports: true
      includeEnums: true
      includeInputs: true
      includeDirectives: true
      includeScalars: true
```

Any help or advice would be greatly appreciated!! 🤗

**Comment by JoviDeCroock** [comment, marked as answer]
Hey, sorry for the delay. Looks like `DataFields` makes it hard to satisfy the requirements on the type in #1675 the `extends` for `UpdateResolver` has been relaxed which should solve your issue. I'm curious if you're seeing this error on other types of resolvers/... In the whole testing of the type-generation I hadn't encountered this yet 😅

**Reply by disbelief** [reply]
> I'm curious if you're seeing this error on other types of resolvers/

Yes, I am. When I comment out the `updates` line, I see similar errors for `resolvers`.
```typescript
export type GraphCacheConfig = {
  // updates?: GraphCacheUpdaters,
  keys?: GraphCacheKeysConfig,
  optimistic?: GraphCacheOptimisticUpdaters,
  resolvers?: GraphCacheResolvers
};
```

Let me know if there's anything else you'd like me to check! Thanks :)

**Reply by JoviDeCroock** [reply]
Thanks for getting back to me! I've added a fix for that one as well, saves us having to go back and forth for multiple patch releases 😅

**Reply by disbelief** [reply]
Fantastic. Will grab the latest once that's merged and let you know if it resolves my issue 🤞

**Reply by JoviDeCroock** [reply]
Should be published now

**Reply by disbelief** [reply]
@JoviDeCroock unfortunately the error persists after updating.

Software:
```
urql/core 2.1.2
@urql/exchange-graphcache 4.1.2
@graphql-codegen/typescript-urql-graphcache 1.0.2
```

Client setup:
```typescript
import { GraphCacheConfig } from '__generated__/types';

const client = createClient({
  url: '/graphql',
  requestPolicy: 'cache-first',
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange<GraphCacheConfig>(),
    fetchExchange,
  ],
});
```

Error:
```
Type 'GraphCacheConfig' does not satisfy the constraint 'Partial<GenericCacheExchangeOpts>'.
  Types of property 'updates' are incompatible.
    Type 'GraphCacheUpdaters | undefined' is not assignable to type 'Partial<GenericUpdatesConfig> | undefined'.
      Type 'GraphCacheUpdaters' is not assignable to type 'Partial<GenericUpdatesConfig>'.
        Types of property 'Mutation' are incompatible.
          Type '{ acceptInvite?: UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs> | undefined; ... 41 more ...; updateUser?: UpdateResolver<...> | undefined; } | undefined' is not assignable to type '{ [fieldName: string]: GenericUpdateResolver<DataFields, Variables>; } | undefined'.
            Type '{ acceptInvite?: UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs> | undefined; ... 41 more ...; updateUser?: UpdateResolver<...> | undefined; }' is not assignable to type '{ [fieldName: string]: GenericUpdateResolver<DataFields, Variables>; }'.
              Property 'acceptInvite' is incompatible with index signature.
                Type 'UpdateResolver<{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }, MutationAcceptInviteArgs>' is not assignable to type 'GenericUpdateResolver<DataFields, Variables>'.
                  Types of parameters 'parent' and 'parent' are incompatible.
                    Property 'acceptInvite' is missing in type 'DataFields' but required in type '{ acceptInvite: Maybe<WithTypename<AcceptInvitePayload>>; }'.ts(2344)
```

**Reply by disbelief** [reply]
Not sure if it's of any use, but in the generated types.ts file, I can produce the same error by specifying the parent type for the exported `GraphCacheConfig`:

```typescript
// __generated__/types.ts (at the bottom of the file)

// export type GraphCacheConfig = {
export interface GraphCacheConfig extends Partial<GenericCacheExchangeOpts> {
  updates?: GraphCacheUpdaters;
  keys?: GraphCacheKeysConfig;
  optimistic?: GraphCacheOptimisticUpdaters;
  resolvers?: GraphCacheResolvers;
}

```

<img width="1084" alt="Screen Shot 2021-05-29 at 10 59 48" src="https://user-images.githubusercontent.com/295574/120066144-497a0a80-c09f-11eb-9d6a-e39d61394137.png">

**Reply by JoviDeCroock** [reply]
That's odd, the ts-playground and my repo don't seem to have this.

https://github.com/JoviDeCroock/graphcache-code-generation/blob/main/example/src/pages/_app.tsx

Playground in the Pr

**Reply by disbelief** [reply]
Sorry which PR has the playground?

I'd be happy to share my repo with you, if you think it might be helpful?

**Reply by disbelief** [reply]
Ah I found the [playground](https://www.typescriptlang.org/play?#code/PTAEAkFMBsAdIE6gC4E94GcBQkAesB7BZFdSUAOQFdpoBBBBAQ1QB4AVAPlAF5QHmbAHY1ooAD6h2EyqIEsOnTgG4sOfERIBLIckQAzJgGNyAMWYBzALaRdGUAG8soUAG19lm7opMbALlAMZAQdCwBdAIA3Ai0AE1UAXzUQUABlIyZoJiQ0TFBs8iECEltkLTLIeyZ7WGySAn1QBErCIQxyWKZkJnVCYlJ4UAAFEKtyrUjyPhFaGRErACNEGQWCAmhIJiEZIJChC1VezVAdPQRDEzSMrIQAeQWAK0gjEicXIwI24KoXogB+AKmKhCF5aT6qFyuADWkFQAV2oQi+SEqESajwfRIuXI6Uy2V4w1G40mMlxN3uTxehwxx1OBmMONQQUgVlMWhgsXsb1AAH0edihL5IPDgqEIby4gDAqL9nMqItlpIZtBxSdYlKEbKlfKlkhtbQ0Ud+tjQABRXTlVBsjkE4GxSD6HSQWKk674yTUWjyNhk7IqI1YsigAAiXSY1ugLr4vr1IbDMk99EYChj3A9cmTbFD3X9NONQezTAJqSZelZ7Mj9gAZHHuhHOdSNPnBoX6-Y+NzobCRXtwgFWxX4lgkgGBuQAKqwTp6ABKlXWkwQrCGBV0hdAeD0Qk5yNQBIHHIw3D4AApnKBas1dAEV1fkIWADRYACUvG40TihxNdCMJlgyAASSESJykgFdUGgAgmCjRxzziHsxXPPkBSFAIAHIgJAvQ0MNfRgVBT4UEqZBFBPXAAgcUAqCnLpEACSdp0gOcMAXRBFFABJXwcEc9CCVgfz-QDgNA8DIOgzgT25ajGIQAIT0vUpXx4bgFN0AA6ZCyEFGxh2fQ4Pi+KiaLOejjKY+doEXAl5NXZAlO4blVOQYc1CAA)

**Reply by JoviDeCroock** [reply]
Tried it out in my own repo with your particular scenario 
![image](https://user-images.githubusercontent.com/17125876/120067992-bd57eb00-c07e-11eb-92ad-af019dfcbb05.png)

**Reply by disbelief** [reply]
I created a gist with my full types.ts here: https://gist.github.com/disbelief/3a105edb47d52dcb0d5e31cfb784970c

(warning: it's large)

**Reply by JoviDeCroock** [reply]
I see, this is an issue when the compiler is in strict-mode which my app kept turning off 😅 hmm not quite sure yet how to solve this.

Most things that would fix this issue would impact the functionality without generics 😅

**Reply by disbelief** [reply]
To be honest, I'm just glad you're able to make any sense of that error message 😲

**Reply by aldrio** [reply]
This specifically comes from `strictFunctionTypes` being enabled. We can get around it by using the [bivarianceHack trick](https://stackoverflow.com/questions/52667959/what-is-the-purpose-of-bivariancehack-in-typescript-types) with the resolver types

[typescript playground showing this](https://www.typescriptlang.org/play?strict=true#code/PTAEAkFMBsAdIE6gC4E94GcBQkAesB7BZFdSUAOQFdpoBBBBAQ1QB4AVAPlAF5QHmbAHY1ooAD6h2EyqIEsOnTgG4sOfERIBLIckQAzJgGNyAMWYBzALaRdGUAG8soUAG19lm7opMbALlAMZAQdCwBdAIA3Ai0AE1UAXzUQUABlIyZoJiQ0TFBs8iECEltkLTLIeyZ7WGySAn1QBErCIQxyWKZkJnVCYlJ4UAAFEKtyrUjyPhFaGRErACNEGQWCAmhIJiEZIJChC1VezVAdPQRDEzSMrIQAeQWAK0gjEicXIwI24KoXogB+AKmKhCF5aT6qFyuADWkFQAV2oQi+SEqESajwfRIuXI6Uy2V4w1G40mMlxN3uTxehwxx1OBmMONQQUgVlMWhgsXsb1AAH0edihL5IPDgqEIby4gDAqL9nMqItlpIZtBxSdYlKEbKlfKlkhtbQ0Ud+tjQABRXTlVBsjkE4GxSD6HSQWKk674yTUWjyNhk7IqI1YsigAAiXSY1ugLr4vr1IbDMk99EYChj3A9cmTbFD3X9NONQezTAJqSZelZ7Mj9gAZHHuhHOdSNPnBoX6-Y+NzobCRXtwgFWxX4lgklgUrcADLB0AAJVNqQnADVTdPQFQMKEUAALLT2T7kADuWlmGTX5BNiAQRGwKTzgcGAFVYJ09NPKutJghWEMCrpC6A8HoQicsiqAEgOHIYNwfAABSjmALi1M0ugBN+SHIIWAA0cGgAAlLw3DRHEqgpBQpoAOoznOi7Lqu66ygsEzZFoWwmOAxhQhhW47qAjq4JUAzkBeV4BgJoCPs+kCvhg76IF+P7ofGAG2MB4GVmBYZtlBjjOKADGRExLGQGxRhQrBCHySh8mYVgOFRDEsQJK4ABEekGSCRnsU5YSNpiokAMKfI6FhUMwZSfAS3JUE+XSIAE4kxVJMkIFhI76MCoLhXoQSsH5-64IBwEBUIQUhV0YJCJw0FGAEfl4Q4I4kQQoAWLYiBaEYGB+FgWXINBkXRWcATQYhpR4Tw3Dci4KTsGQGBGCEsAkFCRT7jU8knPYqmcjp5loQAdHyApCjpSQJDhhzYQA4q1IQdTxRCgFYRDkBg8BGFojpGKuA2IN1QZ0EYJiLQAkkIkTlJA36oNABBMFG2kuHEPZijph1kIK-igAA5KD4N6Fjho9aw-USQgcU-Yl0AfqwANA8guMQ1DMNw5ww6VSTMVk6Aw3yWNE07aA02zfNWiLaAy0EKtoAjboG38IDkAg2DjMsMzsQCzLyAHfy6PHS4p3nWoWBAA) 

This approach seems to be working for me locally, I'm happy to make a PR

**Reply by disbelief** [reply]
@aldrio a PR would be fantastic @JoviDeCroock do you concur? This does seem to be a solution. 

...And today I learned something new about TS 😂

**Reply by JoviDeCroock** [reply]
I've tried applying the `bivarianceHack` you mentioned but to no avail 😅 https://github.com/FormidableLabs/urql/pull/1687

**Reply by aldrio** [reply]
@JoviDeCroock I added your type test into my local branch that I was using before and got it working. I was originally attempting to fix the tests on #1687 but some of the other type changes in that pr affect extra stuff and I don't wanna make too many assumptions since I'm new with urql.

https://github.com/aldrio/urql/tree/bivariance

The main snag here is that GCG will use `null` for arguments in the `GraphCacheResolver` when there's no arguments. I think that’s a bug on GCG’s side, Urql would actually give an empty object in that case, right?

**Reply by JoviDeCroock** [reply]
Yes we should default to `{}` hmm, I think we'd need to fix that in `GCG` in that case 😅

**Reply by JoviDeCroock** [reply]
Issue on GCG will be fixed in the next release https://github.com/dotansimha/graphql-code-generator/pull/6153

**Reply by JoviDeCroock** [reply]
Hey @disbelief 

Is there a way you or me could try out this fix for your case? It's published in GCG and the PR from our side is up as well https://github.com/FormidableLabs/urql/pull/1687

**Reply by disbelief** [reply]
@JoviDeCroock sure. Best way would be for me to try installing urql from that GH branch correct?

**Reply by JoviDeCroock** [reply]
@disbelief the artifacts from [here](https://github.com/FormidableLabs/urql/actions/runs/964300392) are probably the easiest way as that GH branch won't have a built version of graphCache unless you explicitly go build it 😅 alternatively you could reflect the changes in the `.d.ts` files if it's not too much hassle.

**Reply by disbelief** [reply]
@JoviDeCroock understood, thanks. I will try that today and let you know how it goes.

**Reply by disbelief** [reply]
✅  My issue appears to be resolved! 

I upgraded `@graphql-codegen/typescript-urql-graphcache` to `1.0.4` and replaced the `@urql/core` and `@urql/exchange-graphcache` libraries with the artifacts from that PR's build.

Thanks very much @JoviDeCroock and @aldrio

**Reply by JoviDeCroock** [reply]
Thank you so much @aldrio for pushing us in to the right direction, you're a hero!

**Reply by JoviDeCroock** [reply]
This has been published, thanks for your patience!
